runtime/ts/keymgmt/webcrypto.ts

```diff
@@ -0,0 +1,276 @@
+            true, ["encrypt", "decrypt", "wrapKey", "unwrapKey"]);
+        return generatedKey.then(key => new WebCryptoSessionKey(key))
+            .then(skey => skey.disposeToWrappedKeyUsing(deviceKey.publicKey()));
+    }
+
+    static getInstance() {
+        return new WebCryptoKeyGenerator();
+    }
+
+    generateAndStoreRecoveryKey(): PromiseLike<RecoveryKey> {
+        return undefined;
+    }
+
+    generateDeviceKey(): PromiseLike<DeviceKey> {
+        const generatedKey: PromiseLike<CryptoKeyPair> = crypto.subtle.generateKey(
+            {
+                hash: {name: "SHA-512"},
+                name: 'RSA-OAEP',
+                modulusLength: 2048,
+                publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
+            },
+            false, ["encrypt", "decrypt", "wrapKey", "unwrapKey"]);
+        return generatedKey.then(key => new WebCryptoDeviceKey(key));
+    }
+
+    importKey(pemKey: string): PromiseLike<PublicKey> {
+        const key = rs.KEYUTIL.getKey(pemKey);
+        const jwk = rs.KEYUTIL.getJWKFromKey(key);
+
+        return crypto.subtle.importKey("jwk",
+            jwk as JsonWebKey,
+            {
+                name: "RSA-OAEP",
+                hash: {name: "SHA-1"}
+            }, true, ["encrypt", "wrapKey"]).then(ikey => new WebCryptoPublicKey(ikey));
+    }
+}
+
+export class WebCryptoMemoryKeyStorage implements KeyStorage {
+    storageMap: Map<string, Key>;
+
+    constructor() {
+        this.storageMap = new Map();
+    }
+
+    find(keyFingerPrint: string): PromiseLike<Key> {
+        return Promise.resolve(this.storageMap.get(keyFingerPrint));
+    }
+
+    async write(key: Key): Promise<string> {
+        const id = await key.fingerprint();
+        this.storageMap.set(id, key);
+        return Promise.resolve(key.fingerprint());
+    }
+
+    static getInstance() {
+        return new WebCryptoMemoryKeyStorage();
+    }
+}
+
+export class WebCryptoKeyIndexedDBStorage implements KeyStorage {
+
+    async runOnStore(fn: (store: ObjectStore<{}, IDBValidKey>) => PromiseLike<IDBValidKey>) {
+        const db = await idb.open('ArcsKeyManagement', 1,
+            upgradeDB => upgradeDB.createObjectStore('ArcsKeyManagementStore',
+                {autoIncrement: true}));
+
+        const tx = db.transaction('ArcsKeyManagementStore', 'readwrite');
+        const store = tx.objectStore('ArcsKeyManagementStore');
+        const result = await fn(store);
+        await tx.complete;
+        db.close();
+        return Promise.resolve(result);
+    }
+
+    find(keyId: string): PromiseLike<Key> {
+        return undefined;
+    }
+
+    async write(key: Key): Promise<string> {
+        if (key instanceof WebCryptoStorableKey) {
+            const skey = key as WebCryptoStorableKey<CryptoKey>;
+            const fingerprint = await key.fingerprint();
+            await this.runOnStore(store => {
+                return store.put({keyFingerPrint: fingerprint, key: skey.storableKey()});
```

Comment:
Add comment documenting the magic here (link to relevant section of spec if you can dig it up again?)